compiler/src/dotty/tools/dotc/transform/Erasure.scala

```diff
@@ -486,15 +487,20 @@ object Erasure {
           fun1.tpe.widen match {
             case mt: MethodType =>
               val outers = outer.args(fun.asInstanceOf[tpd.Tree]) // can't use fun1 here because its type is already erased
-              var args0 = outers ::: args ++ protoArgs(pt)
+              var args0 = protoArgs(pt, tree.typeOpt)
+              if (mt.paramNames.nonEmpty && !mt.isUnusedMethod) args0 = args ::: args0
+              args0 = outers ::: args0
+
               if (args0.length > MaxImplementedFunctionArity && mt.paramInfos.length == 1) {
                 val bunchedArgs = untpd.JavaSeqLiteral(args0, TypeTree(defn.ObjectType))
                   .withType(defn.ArrayOf(defn.ObjectType))
                 args0 = bunchedArgs :: Nil
               }
               // Arguments are phantom if an only if the parameters are phantom, guaranteed by the separation of type lattices
-              val args1 = args0.filterConserve(arg => !wasPhantom(arg.typeOpt)).zipWithConserve(mt.paramInfos)(typedExpr)
-              untpd.cpy.Apply(tree)(fun1, args1) withType mt.resultType
+              val args1 = args0.filterConserve(arg => !wasPhantom(arg.typeOpt))
+              assert(args1 hasSameLengthAs mt.paramInfos)
```

Comment:
`zipWithConserve` starts with `if (xs.isEmpty || ys.isEmpty) Nil` and therefore it will trim the longest list. Maybe we should change it to `xs.isEmpty && ys.isEmpty` and require the two list to be of the same size.

Reply:
Note that both `x.zip(y)` and `(x, y).zipped` trim the longest list

Reply:
We could change the semantics of ours to ensure no unexpected trimming is done.

Reply:
No, it's fine. I forgot about the trimming behavior. So the current implementation is OK.